File: PROJECT/audio_mirror.py
import os
import time
import shutil
import unicodedata

sanitization_count = 0

# Sanitize a file name
def sanitize_filename(filename):

    # To ensure global variable is referenced
    global sanitization_count

    # Remove any non-ASCII characters and replace wide characters with their closest equivalent
    sanitized_filename = ''.join((c if unicodedata.category(c)[0] not in ('C', 'M', 'Z') else ' ') for c in filename)
    sanitized_filename = unicodedata.normalize('NFKD', sanitized_filename).encode('ASCII', 'ignore').decode('ASCII')

    # If change made
    if filename != sanitized_filename:

        # Increment count
        sanitization_count += 1

        # Print the original filename
        # print("Sanitized:", sanitized_filename)

    return sanitized_filename


# Create folders
def create_directory_structure(source_folder, dest_folder):
    for root, dirs, files in os.walk(source_folder):
        for directory in dirs:
            directory_path = os.path.join(root, directory)
            relative_path = os.path.relpath(directory_path, source_folder)
            new_directory_path = os.path.join(dest_folder, relative_path)
            os.makedirs(new_directory_path, exist_ok=True)


# Create files
def mirror_mp3_files(source_folder, dest_folder):

    file_count = 0
    non_mp3_files = []

    for root, dirs, files in os.walk(source_folder):
        for file in files:

            # If file is an MP3 file
            if file.lower().endswith(".mp3"):
                file_count += 1

                # Sanitize the file name
                sanitized_filename = sanitize_filename(file)

                # Get file path
                file_path = os.path.join(root, sanitized_filename)
                relative_path = os.path.relpath(file_path, source_folder)
                new_file_path = os.path.join(dest_folder, relative_path)
                new_file_path = os.path.splitext(new_file_path)[0] + ".txt"

                # Create an empty .txt file if it doesn't exist already
                if not os.path.exists(new_file_path):
                    with open(new_file_path, 'w') as txt_file:
                        pass

                # Rename the .txt file with the sanitized name using shutil.move
                sanitized_txt_file_path = os.path.join(dest_folder, relative_path)
                sanitized_txt_file_path = os.path.splitext(sanitized_txt_file_path)[0] + ".txt"
                shutil.move(new_file_path, sanitized_txt_file_path)

            else:
                non_mp3_files.append(file)

    return file_count, non_mp3_files


# Create audio mirror
def create_mirror(root_folder, destination_folder):

    # Remove the destination folder if it exists
    if os.path.exists(destination_folder):
        shutil.rmtree(destination_folder)  

    create_directory_structure(root_folder, destination_folder)
    file_count, non_mp3_files = mirror_mp3_files(root_folder, destination_folder)

    return file_count, non_mp3_files


# MAIN
if __name__ == "__main__":

    # Start msg 
    print("\n### Audio Mirror ###")


    ### START ALGORITHM
    start_time = time.time()
    
    # Audio folder path
    folder_path = r'C:\Users\David\Audio'

    # Place mirror in folder next to script
    script_directory = os.path.dirname(os.path.abspath(__file__))
    destination_path = os.path.join(script_directory, "AUDIO_MIRROR")

    # Create mirror
    file_count, non_mp3_files = create_mirror(folder_path, destination_path)

    # Calculate execution time
    execution_time = time.time() - start_time


    ### PRINT INFO

    # Print file count
    print(f"\nMP3 file count: {file_count}")

    # Print sanitization count
    print(f"\nMP3 filenames sanitized: {sanitization_count}")

    # Print non MP3 files found
    print("\nNon-MP3 files found:")
    for file in non_mp3_files:
        print("   " + file)

    # Print time taken
    print(f"\nTime taken: {execution_time:.2f} seconds")

    # Finish msg
    print("\nFinished!\n")

File: PROJECT/AudioMirror/AudioMirror.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Text;

namespace AudioMirror
{
    /// <summary>
    /// Creates a lightweight 'mirror' of an Audio track folder
    /// </summary>
    internal class AudioMirror
    {
        // Audio folder path
        string audioFolderPath = @"C:\Users\David\Audio\";

        // Relative mirror path
        string relMirrorPath = "..\\..\\..\\..\\AUDIO_MIRROR";

        // Filenames sanitized
        private int sanitizationCount = 0;

        // Invalid file name characters
        char[] invalidChars = Path.GetInvalidFileNameChars();


        /// <summary>
        /// Constructor
        /// </summary>
        public AudioMirror()
        {
            // Initialize count
            sanitizationCount = 0;

            // Start
            Console.WriteLine("\n### Audio Mirror ###");
            var start_time = DateTime.Now;

            // Set destination mirror path relative to where program is running
            string programDir = AppDomain.CurrentDomain.BaseDirectory;
            string mirrorPath = Path.Combine(programDir, relMirrorPath);

            // Setup folders
            SetupFolders(audioFolderPath, mirrorPath);

            // Populate with files
            var mirrorInfo = MirrorMp3Files(audioFolderPath, mirrorPath);

            // Calculate execution time
            var executionTime = DateTime.Now - start_time;

            // Print info
            PrintInfo(mirrorInfo, executionTime);
        }


        /// <summary>
        /// Set up folder structure
        /// </summary>
        /// <param name="srcFolder">The folder being mirrored</param>
        /// <param name="destFolder">The destination folder</param>
        /// <returns></returns>
        public void SetupFolders(string srcFolder, string destFolder)
        {
            // Remove the destination folder if it exists
            if (Directory.Exists(destFolder))
            {
                Directory.Delete(destFolder, true);
            }

            // Get sub dirs
            string[] dirs = Directory.GetDirectories(srcFolder, "*", SearchOption.AllDirectories);

            // For each directory
            foreach (var directoryPath in dirs)
            {
                // Create equivalent directory
                string relativePath = GetRelativePath(srcFolder, directoryPath);
                string newDirectoryPath = Path.Combine(destFolder, relativePath);
                Directory.CreateDirectory(newDirectoryPath);
            }
        }


        /// <summary>
        /// Populate folder structure with mirrored files
        /// </summary>
        /// <param name="srcFolder">The folder being mirrored</param>
        /// <param name="destFolder">The destination folder</param>
        /// <returns></returns>
        private Tuple<int, List<string>> MirrorMp3Files(string srcFolder, string destFolder)
        {
            // Holders
            int fileCount = 0;
            List<string> nonMp3Files = new List<string>();

            // For every file
            foreach (var filePath in Directory.GetFiles(srcFolder, "*", SearchOption.AllDirectories))
            {
                // If its an MP3 file
                if (Path.GetExtension(filePath)?.ToLower() == ".mp3")
                {
                    // Get relative file path
                    string relativePath = GetRelativePath(srcFolder, filePath);

                    // Replace file name part with sanitized file name
                    string fileName = Path.GetFileName(filePath);
                    string sanitizedFilename = SanitizeFilename(fileName);
                    relativePath = relativePath.Replace(fileName, sanitizedFilename);

                    // Get full path
                    string newFilePath = Path.Combine(destFolder, relativePath);
                    newFilePath = Path.ChangeExtension(newFilePath, ".txt");

                    // Create an empty .txt file if it doesn't exist already
                    if (!File.Exists(newFilePath))
                    {
                        File.WriteAllText(newFilePath, string.Empty);
                    }

                    // Increment file count
                    fileCount++;
                }
                else
                {
                    // Else if it is not an MP3 file, add to list
                    nonMp3Files.Add(GetRelativePath(srcFolder, filePath));
                }
            }

            // Return holders
            return Tuple.Create(fileCount, nonMp3Files);
        }


        /// <summary>
        /// Print info about completed mirroring process 
        /// </summary>
        /// <param name="mirrorInfo"></param>
        /// <param name="executionTime"></param>
        private void PrintInfo(Tuple<int, List<string>> mirrorInfo, TimeSpan executionTime)
        {
            // Extract mirror info items
            int fileCount = mirrorInfo.Item1;
            List<String> nonMp3Files = mirrorInfo.Item2;

            // Print file count
            Console.WriteLine($"\nMP3 file count: {fileCount}");

            // Print sanitization count
            Console.WriteLine($"\nMP3 filenames sanitized: {sanitizationCount}");

            // Print non-MP3 file info
            Console.WriteLine($"\nNon-MP3 files found: {nonMp3Files.Count}");
            //foreach (var file in nonMp3Files)
            //{
            //    Console.WriteLine("   " + file);
            //}

            // Print time taken
            Console.WriteLine($"\nTime taken: {Math.Round(executionTime.TotalSeconds, 3)} seconds");

            // Finish message
            Console.WriteLine("\nFinished!\n");
        }


        /// <summary>
        /// Sanitize a given filename of special characters
        /// </summary>
        /// <param name="filename">Original filename.</param>
        /// <returns>Sanitized filename.</returns>
        private string SanitizeFilename(string filename)
        {
            // Remove any non-ASCII characters and replace wide characters with their closest equivalent
            string sanitizedFilename = new string(filename
                .Where(c => Char.GetUnicodeCategory(c) != UnicodeCategory.Control &&
                            Char.GetUnicodeCategory(c) != UnicodeCategory.ModifierSymbol &&
                            Char.GetUnicodeCategory(c) != UnicodeCategory.OtherSymbol)
                .ToArray());
            sanitizedFilename = Encoding.ASCII.GetString(Encoding.GetEncoding("Cyrillic").GetBytes(sanitizedFilename));

            // Replace invalid characters with an underscore if found
            if (sanitizedFilename.Any(c => invalidChars.Contains(c)))
            {
                sanitizedFilename = new string(sanitizedFilename.Select(c => invalidChars.Contains(c) ? '_' : c).ToArray());
            }

            // If filename was sanitized
            if (filename != sanitizedFilename)
            {
                // Increment count
                sanitizationCount++;

                // Print the filenames
                //Console.WriteLine("\nOriginal: " + filename);
                //Console.WriteLine("Sanitized: " + sanitizedFilename);
            }

            return sanitizedFilename;
        }


        /// <summary>
        /// Calculates the relative path from a base path to a target path.
        /// </summary>
        /// <param name="basePath">The base path.</param>
        /// <param name="targetPath">The target path.</param>
        /// <returns>The relative path from the base path to the target path.</returns>
        private string GetRelativePath(string basePath, string targetPath)
        {
            // Convert paths to URIs for accurate relative path calculation
            Uri baseUri = new Uri(basePath);
            Uri targetUri = new Uri(targetPath);

            // Calculate relative URI
            Uri relativeUri = baseUri.MakeRelativeUri(targetUri);

            // Convert relative URI to a string and unescape special characters
            return Uri.UnescapeDataString(relativeUri.ToString());
        }
    }
}
